common/network-common/src/main/java/org/apache/spark/network/sasl/SparkSaslAES.java

```diff
@@ -0,0 +1,252 @@
+
+import com.intel.chimera.cipher.Cipher;
+import com.intel.chimera.cipher.CipherTransformation;
+import com.intel.chimera.utils.Utils;
+
+import javax.crypto.Mac;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
+import javax.security.sasl.SaslException;
+import java.io.IOException;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+import java.util.Properties;
+
+/**
+ * AES encryption and decryption.
+ */
+public class SparkSaslAES {
+  private static final byte[] EMPTY_BYTE_ARRAY = new byte[0];
+
+  private final Cipher encryptor;
+  private final Cipher decryptor;
+
+  private final Integrity integrity;
+
+  public SparkSaslAES(CipherTransformation cipherTransformation, Properties properties, byte[] inKey,
+      byte[] outKey, byte[] inIv, byte[] outIv) throws IOException {
+    checkTransformation(cipherTransformation);
+    // encryptor
+    encryptor = Utils.getCipherInstance(cipherTransformation, properties);
+    try {
+      encryptor.init(Cipher.ENCRYPT_MODE, outKey, outIv);
+    } catch (InvalidKeyException | InvalidAlgorithmParameterException e) {
+      throw new IOException("Failed to initialize encryptor", e);
+    }
+
+    // decryptor
+    decryptor = Utils.getCipherInstance(cipherTransformation, properties);
+    try {
+      decryptor.init(Cipher.DECRYPT_MODE, inKey, inIv);
+    } catch (InvalidKeyException | InvalidAlgorithmParameterException e) {
+      throw new IOException("Failed to initialize decryptor", e);
+    }
+
+    integrity = new Integrity(outKey, inKey);
+  }
+
+  /**
+   * Encrypts input data. The result composes of (msg, padding if needed, mac) and sequence num.
+   * @param data the input byte array
+   * @param offset the offset in input where the input starts
+   * @param len the input length
+   * @return the new encrypted byte array.
+   * @throws SaslException if error happens
+   */
+  public byte[] wrap(byte[] data, int offset, int len) throws SaslException {
+    // mac
+    byte[] mac = integrity.getHMAC(data, offset, len);
+    integrity.incMySeqNum();
+
+    // padding based on cipher
+    byte[] padding;
+    if (encryptor.getTransformation() == CipherTransformation.AES_CBC_NOPADDING) {
+      int bs = encryptor.getTransformation().getAlgorithmBlockSize();
+      int pad = bs - (len + 10) % bs;
+      padding = new byte[pad];
+      for (int i = 0; i < pad; i ++) {
+        padding[i] = (byte) pad;
+      }
+    } else {
+      padding = EMPTY_BYTE_ARRAY;
+    }
+
+    // encrypt
+    byte[] toBeEncrypted = new byte[len + 10 + padding.length];
+    System.arraycopy(data, offset, toBeEncrypted, 0, len);
+    System.arraycopy(padding, 0, toBeEncrypted, len, padding.length);
+    System.arraycopy(mac, 0, toBeEncrypted, len + padding.length, 10);
+    byte[] encrypted = encryptor.update(toBeEncrypted, 0, toBeEncrypted.length);
```

Comment:
Why not call encryptor.update() for data, paddding and mac separately? It can save the unnecessary copy from data, padding and mac to toBeEncrypted.